invenio_preservation_sync/api.py

```diff
@@ -0,0 +1,141 @@
+    """
+
+    @classmethod
+    def create(
+        cls,
+        record_id,
+        revision_id,
+        status,
+        harvest_timestamp=None,
+        archive_timestamp=None,
+        uri=None,
+        path=None,
+        event_id=None,
+        description=None,
+    ):
+        """Create a Preservation Information."""
+        status = PreservationInfoAPI._convert_status(status)
+        preservation = PreservationInfo(
+            record_id=record_id,
+            revision_id=revision_id,
+            status=status,
+            harvest_timestamp=harvest_timestamp,
+            archive_timestamp=archive_timestamp,
+            uri=uri,
+            path=path,
+            event_id=event_id,
+            description=description,
+        )
+
+        db.session.add(preservation)
+        db.session.commit()
+        return preservation
+
+    @classmethod
+    def get_by_record_id(cls, record_id=None, latest=False):
+        """Get Preservation Info by record id."""
+        preservation = PreservationInfo.query.filter_by(record_id=record_id).order_by(
+            PreservationInfo.revision_id.desc(),
+            PreservationInfo.archive_timestamp.desc(),
+            PreservationInfo.created.desc(),
+        )
+        if latest:
+            return preservation.first()
+        return preservation.all()
+
+    @classmethod
+    def get_existing_preservation(
+        cls,
+        record_id=None,
+        revision_id=None,
+        archive_timestamp=None,
+    ):
+        """Return Preservation Info if it already exists."""
+        preservation = PreservationInfo.query.filter_by(
+            record_id=record_id,
+            revision_id=revision_id,
+            archive_timestamp=archive_timestamp,
+        ).first()
+        return preservation
+
+    @classmethod
+    def update_existing_preservation(
+        cls,
+        preservation,
+        status=None,
+        harvest_timestamp=None,
+        uri=None,
+        path=None,
+        event_id=None,
+        description=None,
+    ):
+        """Update existing preservation."""
+        status = PreservationInfoAPI._convert_status(status)
+
+        if (
+            preservation.status == status
+            and preservation.harvest_timestamp == parse(harvest_timestamp)
+            and preservation.uri == uri
+            and preservation.path == path
+            and preservation.description == description
+        ):
+            raise PreservationAlreadyReceivedError(preservation)
+
+        if status:
+            preservation.status = status
+
+        if harvest_timestamp:
+            preservation.harvest_timestamp = harvest_timestamp
+
+        if uri:
+            preservation.uri = uri
+
+        if path:
+            preservation.path = path
+
+        if description:
+            preservation.description = description
+
+        preservation.event_id = event_id
+
+        db.session.commit()
+
+    @classmethod
+    def _convert_status(cls, value):
```

Comment:
I would have moved this inside the `PreservationStatus` enum class, so the conversion str -> obj can be re-used everywhere, not only inside here.